basic tutorials (#1)
keypom/keypom-docs: docs/Tutorials/Basics/fc-drops.md

<TabItem value="Mac/Lnx" label="Mac OS/Linux">

```bash
touch fc-keypom.js
```

</TabItem>
<TabItem value="WNDS" label="Windows">

```bash
fsutil file createnew fc-keypom.js 0
```

</TabItem>
</Tabs>

Finally, the last step is to install the Keypom JS SDK.
```bash
npm install keypom-js
```

After installing the SDK, your `package.json` file should now look slightly different.

<details>
<summary>package.json after installing the SDK</summary>
<p>

```bash
{
  "name": "my-keypom-project",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  # highlight-start
  "dependencies": {
    "keypom-js": "^1.2.0-rc.1"
  }
  # highlight-end
}
```

</p>
</details>

With these steps complete, your project folder should look like this. 

```bash
/my-keypom-project
├── fc-keypom.js
├── package.json
├── package-lock.json
├── node_modules
│   └── keypom-js
│   └── ...
```

You are now ready to begin creating your drop!

---

## Breaking Down the Problem

A Function Call Drop is Keypom's most powerful drop type. A brief breakdown of how an FC drop works is as follows:

> :::info

Reviewer: I would remove this as an info tag

For every key-use, a vector of functions is called in the order that they are defined. This vector can vary for each individual claim.

This means for each key use, a different set of functions can be called. An example can be seen below. 
:::

| Key Use    | Functions called during n<sup>th</sup> Key Use |
| -----------| ---------------------------------------------- |
| Key Use 1  | `nft_mint`                                     |
| Key Use 2  | `my_nft_function_1`, `my_nft_function_2`       |
| Key Use 3  | `nft_transfer_call`                            |

:::note

> In this tutorial, the key will only call `nft_mint`.

Reviewer: Remove this as a note and replace with: 
```suggestion
In this tutorial, the key will be single-use and only call `nft_mint`.
```

:::

> With the power of function call drops understood, the process of creating one can be broken down.

Reviewer: ```suggestion
The process of creating an FC drop is similar to the other drop types:
```

}

fcDropKeypom()

```
## Getting Started
In this section, you'll be addressing the first step: connecting to NEAR. 

This is done with `NEAR-API-JS` and consists of:

1) Create a Keystore, which stores your access keys used to sign transactions   
  * select a network, either `testnet` or `mainnet`  
  * choose a location where the keypairs live, either a folder on your local machine, or in-memory      

2) Define a NEAR configuration using the Keystore  
3) Use the configuration to initialize a connection to NEAR  

More information about this process can be found [here](https://docs.near.org/tools/near-api-js/quick-reference#key-store).

:::note
For simplicity, this tutorial will choose a file-based keystore and point to the `~/.near-credentials` folder on your local machine since this is where most of your keys are stored. For more information about KeyStores, visit NEAR's [official docs](https://docs.near.org/tools/near-api-js/quick-reference#key-store).
:::

```js reference
https://github.com/keypom/keypom-js/blob/bbe4716ff64dd7a73a6d727a5aea518e8141f60f/docs-examples/keypom-js-sdk/fc-example.js#L8-L25
```

## Creating Drop with Function Call Data

> In this section, you'll learn about the process of creating a FC drop using the SDK. 

Reviewer: ```suggestion
In this section, you'll learn about the process of creating an FC drop using the SDK. 
```

> This process starts with calling the `initKeypom` function. This will always be the first function you call to interact with the SDK. 

Reviewer: ```suggestion
This starts with calling the `initKeypom` function and will always be the first function you call to interact with the SDK. 
```

`initKeypom` initializes the SDK to allow for interactions with the Keypom smart contracts. Without it, none of the other SDK functions would work as expected. If a NEAR connection is not already present, it will initialize a new one for you. More info on the `initKeypom` function can be found [here](../../keypom-sdk/modules#initkeypom).

After `initKeypom` is called, the FC Drop can be created by calling `createDrop` and adding an `fcData` parameter. 

:::tip
Recall that the private keys being generated using `createDrop` are used to store the assets. These keys are then embedded within a link.

In an FC Drop, the assets consist of a set of invokable methods and optional $NEAR.
:::

The primary task in creating the Function Call Drop is to define `fcData`. It is an object with the following properties.

```bash
fcData

> ├── FCConfig

Reviewer: ```suggestion
├── FCConfig?
```

├── methods
```

> - `FCConfig`: [Specific configuratrions](../../keypom-sdk/interfaces/FCConfig.md) for the FC Drop.  

Reviewer: ```suggestion
- `FCConfig`: [Specific configurations](../../keypom-sdk/interfaces/FCConfig.md) for the FC Drop.  
```

<p align="center"> <img src={require("/static/img/docs/basic-tutorials/fc/explorer.png").default} alt="explorer transactions" width="80%"/> </p>

Within the [`claim`](https://explorer.testnet.near.org/transactions/Cgckyr9EEC3o9MDiL8E2mvs6JYPuCApYqWjpgZvy3z6U) transaction, you can also see that `nft_mint` was called on the `nft.examples.testnet` contract. 
<p align="center"> <img src={require("/static/img/docs/basic-tutorials/fc/nft-mint.png").default} alt="explorer transactions" width="80%"/> </p>

This can be confirmed by visiting the "Collectibles" tab in your NEAR wallet. You should see the newly minted NFT in your wallet.  
<p align="center"> <img src={require("/static/img/docs/basic-tutorials/nft/collectibles-claimed2.png").default} alt="NFT collectibles claim" width="65%"/> </p>

---

## Conclusion
In this tutorial, you learned the how to [create a function call drop](fc-drops.md#creating-drop-with-function-call-data) using the `fcData` parameter. 

> Now that you've had a good introduction to creating all 4 Keypom drop types, you can tinker with existing code in the [Deploy Scripts](getting-started.md#deploy-scripts) or move on to the [Advanced Tutorials](../Advanced/ticketing/concept.md) for more challenging and practical examples.

Reviewer: ```suggestion
Now that you've had a good introduction to creating all 4 Keypom drop types, feel free to modify the scripts created or move on to the [Advanced Tutorials](../Advanced/ticketing/concept.md) for more challenging and practical examples.
```
